### ISSUE TITLE
Impossible to extract features in Multi Animal Tracking

### ISSUE THREAD

**Issue by Lucas97223** [sgoldenlab/simba#236]
**Describe the bug**
I'm working with multi animal tracking from DLC from h5 files

**To Reproduce**
1. Go to Load Project and load one
2. Click on the Extract Features tab
3. Click on Extract features
4. See error : 
"Exception in thread Thread-72:
Traceback (most recent call last):
  File "C:\Users\Admin\anaconda3\envs\S\lib\threading.py", line 980, in _bootstrap_inner
    self.run()
  File "C:\Users\Admin\anaconda3\envs\S\lib\threading.py", line 917, in run
    self._target(*self._args, **self._kwargs)
  File "C:\Users\Admin\anaconda3\envs\S\lib\site-packages\simba\SimBA.py", line 2258, in extractfeatures
    feature_extractor.extract_features()
  File "C:\Users\Admin\anaconda3\envs\S\lib\site-packages\simba\features_scripts\feature_extractor_16bp.py", line 138, in extract_features
    self.out_data['Mouse_1_poly_area'] = Parallel(n_jobs=-1, verbose=0, backend="threading")(delayed(self.convex_hull_calculator_mp)(x, self.px_per_mm) for x in mouse_1_ar)
  File "C:\Users\Admin\anaconda3\envs\S\lib\site-packages\joblib\parallel.py", line 1098, in __call__
    self.retrieve()
  File "C:\Users\Admin\anaconda3\envs\S\lib\site-packages\joblib\parallel.py", line 975, in retrieve
    self._output.extend(job.get(timeout=self.timeout))
  File "C:\Users\Admin\anaconda3\envs\S\lib\multiprocessing\pool.py", line 771, in get
    raise self._value
  File "C:\Users\Admin\anaconda3\envs\S\lib\multiprocessing\pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "C:\Users\Admin\anaconda3\envs\S\lib\site-packages\joblib\_parallel_backends.py", line 620, in __call__
    return self.func(*args, **kwargs)
  File "C:\Users\Admin\anaconda3\envs\S\lib\site-packages\joblib\parallel.py", line 288, in __call__
    return [func(*args, **kwargs)
  File "C:\Users\Admin\anaconda3\envs\S\lib\site-packages\joblib\parallel.py", line 288, in <listcomp>
    return [func(*args, **kwargs)
  File "C:\Users\Admin\anaconda3\envs\S\lib\site-packages\simba\features_scripts\feature_extractor_16bp.py", line 99, in convex_hull_calculator_mp
    return ConvexHull(arr).area / px_per_mm
  File "_qhull.pyx", line 2458, in scipy.spatial._qhull.ConvexHull.__init__
  File "_qhull.pyx", line 353, in scipy.spatial._qhull._Qhull.__init__
scipy.spatial._qhull.QhullError: QH6154 Qhull precision error: Initial simplex is flat (facet 1 is coplanar with the interior point)

While executing:  | qhull i Qt
Options selected for Qhull 2019.1.r 2019/06/21:
  run-id 1840749349  incidence  Qtriangulate  _pre-merge  _zero-centrum
  _max-width 4.2e+02  Error-roundoff 3.4e-13  _one-merge 1.7e-12
  _near-inside 8.4e-12  Visible-distance 6.7e-13  U-max-coplanar 6.7e-13
  Width-outside 1.3e-12  _wide-facet 4e-12  _maxoutside 2e-12

The input to qhull appears to be less than 2 dimensional, or a
computation has overflowed.

Qhull could not construct a clearly convex simplex from points:
- p6(v3): 1.8e-25 4.2e-25
- p7(v2): 1.2e+02 4.2e+02
- p0(v1): 3e-27 5.6e-27

The center point is coplanar with a facet, or a vertex is coplanar
with a neighboring facet.  The maximum round off error for
computing distances is 3.4e-13.  The center point, facets and distances
to the center point are as follows:

center point    39.32    140.7

facet p7 p0 distance= -7.1e-15
facet p6 p0 distance=  -18
facet p6 p7 distance= -7.1e-15

These points either have a maximum or minimum x-coordinate, or
they maximize the determinant for k coordinates.  Trial points
are first selected from points that maximize a coordinate.

The min and max coordinates for each dimension are:
  0:  3.027e-27     117.9  difference= 117.9
  1:  5.558e-27       422  difference=  422

If the input should be full dimensional, you have several options that
may determine an initial simplex:
  - use 'QJ'  to joggle the input and make it full dimensional
  - use 'QbB' to scale the points to the unit cube
  - use 'QR0' to randomly rotate the input for different maximum points
  - use 'Qs'  to search all points for the initial simplex
  - use 'En'  to specify a maximum roundoff error less than 3.4e-13.
  - trace execution with 'T3' to see the determinant for each point.

If the input is lower dimensional:
  - use 'QJ' to joggle the input and make it full dimensional
  - use 'Qbk:0Bk:0' to delete coordinate k from the input.  You should
    pick the coordinate with the least range.  The hull will have the
    correct topology.
  - determine the flat containing the points, rotate the points
    into a coordinate plane, and delete the other coordinates.
  - add one or more points to make the input full dimensional."

**Expected behavior**
I expect to have the features_extracted folder in my project folder, which works when I select Simple Tracking, but i need to do it for Multi Animal Tracking

**Desktop (please complete the following information):**
 - Windows 10
 - Python Version 3.10.10
 - conda version 22.9.0

**Comment by sronilsson** [comment]
Hello @Lucas97223! Which version of SImBA are you working with? `pip show simba-uw-tf-dev`

**Comment by sronilsson** [comment]
There is more information on this error here: however, there was a fix for this error inserted fairly recently:

https://github.com/sgoldenlab/simba/blob/master/docs/FAQ.md#3-i-get-a-qhull-eg-qh6154-or-6013-error-when-extracting-the-features

**Comment by Lucas97223** [comment]
Hi, thank you for your very quick answer. I'm currently using the version 1.50.1 of Simba.
Tahnk you, I'll take a look at the link you sent

**Comment by Lucas97223** [comment]
Hi again,
I read the issue solution from the link and thought about it.
First of all, no body-part locations were filtered in the pose-estimation performed via DLC, so the second option of the solution is not checked in my situation.
Regarding, the first option, the fact that the animals disappear sometimes is very important in my experiments. My arena contains a nest and when the mice enter in it, they disappear and that is a very important information that we don't want to loose. Considering your solution, the only way that I solve the problem is to interpolate the pose estimation data. This is a good idea but the problem is that the mice can spend a lot of time in the nest. Hence, for them not to disappear, I would have to set a very long period for the interpolation and I am afraid it is going to mess with the data when the mice are actually visible by the camera.
Do you have an idea of how I could do to solve this problem?
If not, I think the best thing to do is to run the nest analysis as if it was a classical traking situation, and analyse the other data by removing the moments they are in the nest. This solution still makes me loose a lot of data.
Thank you in advance!

**Comment by sronilsson** [comment]
Got it.

TLDR: can you update to simba `1.52.4` with `pip install simba-uw-tf-dev --upgrade` and see if it works?

The error comes from being unable to calculate the volume of the animals. I **thought** it was because all the body-parts landed on the same coordinate (the animal is one-dimensional), and was surprised, as we should catch 1D events now and set the animal volume to `0`. However I looked back at your error and that does not seem to be it - you seem to have sub-pixel data with body-parts e.g., landing on eg pixel 1.8e-25 and pixel 4.2e-25 and I am guessing [scipy.spatial.ConvexHull](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.ConvexHull.html) struggle with these values. I've inserted a fix where the values are turned to ints before trying to calculate hull.

**Comment by sronilsson** [comment]
Thanks for testing @Lucas97223 - could you drop the CSV  for the video that errors from the `project_folder/csv/outlier_corrected_movement_location` here and I will take a look? If too large then try zip or gdrive link? Thanks!

**Comment by Lucas97223** [comment]
Please fin attached the link for the csv file for the right video:
https://drive.google.com/drive/folders/1NjKU89mUKrBSw_Uly5o762jEtGZz6eOD?usp=sharing

**Comment by sronilsson** [comment]
Thanks for sharing @Lucas97223, very helpful, inserted a fix when scipy.spatial.ConvexHull can't handle the numbers. Will push it later and let you know. PS. The video is more than half a million frames, so took me nearly 30min to extract features from that file on my i5

**Comment by Lucas97223** [comment]
Ok, thank you very much for your help and availability!
I know it is very long (the others will be 12h long, so double of the one I shared).
Have a nice day

**Comment by sronilsson** [comment]
👍🏻  Can you test how it looks after `pip install simba-uw-tf-dev --upgrade` and let me know how it goes?

**Comment by Lucas97223** [comment]
Hi again,
I think this version runs correctly with my data. I did not run the whole process of extracting features but it seems to work. I will do it in two days and let you know: Thank for everything

**Comment by Lucas97223** [comment]
Hi, I just ran the extraction of the features on my Multi Animal Tracking data and it worked! Thank you very much for your help.

**Comment by Lucas97223** [comment]
Hi again,
I just want to signal to changes that I spotted using the aforementioned version of simba (1.52.5).
First one that does not seem a problem to me: the graphic interface of this version has changed when loading a project in the [Further imports] tab.
Second one is an issue for me: since I upgraded simba, the ROI analysis is not performed anymore on the ROI design thanks to the "polygon" shape. The quantifications do not appear in the files created via ANALYZE ROI DATA. Could it be a problem with the new version or is it an other problem independent to that ?
Thanks in advance

**Comment by sronilsson** [comment]
Hi @Lucas97223 - again, thanks for reporting! I tried to recreate it but it worked fine on my end so need a little more info.

Which button is it that you are using, is it the button in the screengrab below? 

About the [further imports] tab: yes, I realized all the buttons and windows where floating around un-aligned and looked a little crazy, they have just but put in their own frame. 


<img width="521" alt="image" src="https://user-images.githubusercontent.com/34761092/225301630-a70d700e-af49-4d95-a0e4-549491c682ce.png">

**Comment by Lucas97223** [comment]
Yeah I use this button, it also doesn't work when I append the ROI features to the other features, but I guess it is a consequence of the first problem.

**Comment by sronilsson** [comment]
What error do you get when you append ROI features?

**Comment by Lucas97223** [comment]
I have no error during the process, either for ROI analysis or for appending ROI data to features_extracted.
I just tried with the updated version but it did not work either.

**Comment by sronilsson** [comment]
Got it, does the project still live in that gdrive link, or do you have a new one you are working with?

**Comment by Lucas97223** [comment]
It is a new one. New project but same video.

**Comment by sronilsson** [comment]
Could you place the new one in the same gdrive and I can see if I can replicate the error on my end with your project?

**Comment by Lucas97223** [comment]
I put the config.ini file in the folder because i don't have enough place to put all the project. Is it enough or shall I found another way of sharing the project?

**Comment by sronilsson** [comment]
Thanks @Lucas97223, but I need more to be able to recreate your error: It's enough with a project with a single file within the `project_folder/csv/outlier_corrected_movement_location` folder. If you have large video files, or any other large files within the sub-directories within the `project/csv` subdirectories, then you can skip sharing those.

**Comment by sronilsson** [comment]
Thanks @Lucas97223 ! It looks like you will have to grant access

**Comment by Lucas97223** [comment]
Sorry for that. I think it is okay now.
Besides, I had another project running with a polygon shape as a test and it has worked well, the ROI analysis is also performed for this shape. Hence, I think that the problem comes from my current project and not with the version. I will just try and re-do the project to see if the problem persists.

**Comment by sronilsson** [comment]
Thanks, let me know!
